webapp/src/main/java/org/opentestsystem/rdw/admin/service/impl/DefaultCsvValidationService.java

```diff
@@ -0,0 +1,200 @@
+        final Set<String> headers = csvHeaders.stream().map(String::toLowerCase).collect(Collectors.toSet());
+        if (!headers.containsAll(RequiredHeaders)) {
+            final Set<String> missing = newHashSet(RequiredHeaders);
+            missing.removeAll(headers);
+            failures.add(CsvValidationFailure.builder()
+                    .row(0)
+                    .message("Missing headers " + missing + ", found " + headers)
+                    .build());
+        }
+        return failures;
+    }
+
+    @Override
+    public List<CsvValidationFailure> validateRecords(final Iterator<CSVRecord> recordIterator,
+                                                      final PermissionScope permissionScope) {
+        final List<CsvValidationFailure> validationFailures = new ArrayList<>();
+
+        final Set<String> availableSchools = schoolRepository.findAll(permissionScope).stream()
+                .map(School::getNaturalId)
+                .collect(Collectors.toSet());
+
+        final Set<String> completedSchools = new HashSet<>();
+        final Set<String> completedSchoolGroupYears = new HashSet<>();
+
+        String currentSchool = null;
+        String currentSchoolGroupYear = null;
+        CsvSubject currentSchoolGroupYearSubject = null;
+
+        while (recordIterator.hasNext()) {
+            final CSVRecord record = recordIterator.next();
+            final String school = record.get(HeaderSchool);
+            final String group = record.get(HeaderGroup);
+            final Integer schoolYear = getAsInt(record, HeaderYear);
+            final CsvSubject subject = CsvSubject.valueOfCaseInsensitive(record.get(HeaderSubject));
+
+            if (isBlank(school) && isBlank(group) && schoolYear == null) {
+                //Not a viable record, skip
+                continue;
+            }
+            validateKeyValues(validationFailures, school, group, schoolYear, record.getRecordNumber());
+            if (!validationFailures.isEmpty()) continue;
+
+            //Validate the School
+            if (currentSchool == null || !currentSchool.equals(school)) {
+                if (currentSchool != null) completedSchools.add(currentSchool);
+                completedSchoolGroupYears.clear();
+                currentSchoolGroupYear = null;
+                currentSchool = school;
+
+                validateSchool(validationFailures, availableSchools, completedSchools, school, record.getRecordNumber());
+            }
+
+            //Validate the School, Group, and Year
+            final String schoolGroupYear = school + group + String.valueOf(schoolYear);
+            if (currentSchoolGroupYear == null || !currentSchoolGroupYear.equalsIgnoreCase(schoolGroupYear)) {
+                if (currentSchoolGroupYear != null) completedSchoolGroupYears.add(currentSchoolGroupYear);
+                currentSchoolGroupYearSubject = null;
+                currentSchoolGroupYear = schoolGroupYear;
```

Comment:
Hm, I sense a pattern of very similar code here, any chance to extract a method for reuse?